### ISSUE TITLE
Changing Mnesia table 'ejabberd_commands' from unknown to ram_copies fails

### ISSUE THREAD

**Issue by joudinet** [processone/ejabberd#2518]
> What version of ejabberd are you using?

17.11

> What operating system (version) are you using?

Buildroot (2018.02.1)

> How did you install ejabberd (source, package, distribution)?

source

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

The command join_cluster fails sometimes with the following error reported in the log:
```
2018-07-09 14:07:22.532 [info] <0.723.0>@ejabberd_mnesia:change_table_copy_type:162
 Changing Mnesia table 'ejabberd_commands' from unknown to ram_copies
2018-07-09 14:07:22.551 [error] <0.723.0>@ejabberd_mnesia:mnesia_op:440
 failure on mnesia change_table_copy_type
 [ejabberd_commands,'ejabberd@10.127.171.160',ram_copies]:
 {aborted,{no_exists,ejabberd_commands}}
2018-07-09 14:07:22.554 [error] <0.725.0>
 Supervisor ejabberd_sup had child ejabberd_commands
 started with ejabberd_commands:start_link()
 at undefined exit with reason {aborted,{no_exists,ejabberd_commands}}
 in context start_error
2018-07-09 14:07:22.554 [error] <0.743.0> CRASH REPORT
 Process <0.743.0> with 0 neighbours exited with reason:
 {aborted,{no_exists,ejabberd_commands}} in mnesia:abort/1 line 350
2018-07-09 14:07:22.626 [critical] <0.682.0>@ejabberd_app:start:63
 Failed to start ejabberd application:
 {error,{shutdown,{failed_to_start_child,ejabberd_commands,
  {aborted,{no_exists,ejabberd_commands}}}}}
```

I believe there is some sort of race conditions here as this command does not always fail.
My setup is three devices running each an instance of ejabberd 17.11, with SPOOL_DIR=/tmp/ejabberd. So they all create a new cluster at every boot.
I noticed this issue by having one the three devices not joining the cluster (i.e., ejabberdctl join_cluster returns false on one of the device, even though it should be able to join the cluster).

**Comment by joudinet** [comment]
Thanks @zinid for the feedback. So, I can assume this bug is still present in ejabberd 18.06?
In the meantime, I tried to applied some commits to my 17.11 build to see if they could fix this issue. In particular, I applied a3083b5d and 548ef7b8. Since then, I haven't seen the issue reported here but I get crashes in the bosh module:
```
2018-07-17 16:27:12.869 [error] <0.402.0>@mod_muc_mnesia:handle_info:343 unexpected info: {mnesia_system_event,{mnesia_up,'ejabberd@10.127.171.160'}}
2018-07-17 16:27:18.985 [error] <0.402.0>@mod_muc_mnesia:handle_info:343 unexpected info: {mnesia_system_event,{mnesia_up,'ejabberd@10.127.171.160'}}
2018-07-17 16:29:36.470 [error] <0.1020.0>@ejabberd_bosh:drop_holding_receiver:701 gen_fsm <0.1020.0> in state active terminated with reason: bad record state in ejabberd_bosh:drop_holding_receiver/1 line 701
2018-07-17 16:29:36.471 [error] <0.1020.0>@ejabberd_bosh:drop_holding_receiver:701 CRASH REPORT Process <0.1020.0> with 0 neighbours exited with reason: bad record state in ejabberd_bosh:drop_holding_receiver/1 line 701 in p1_fsm:terminate/8 line 760
2018-07-17 16:29:36.475 [error] <0.474.0>@ejabberd_bosh:drop_holding_receiver:701 Supervisor 'mod_bosh_gnyoi.com' had child undefined started with {ejabberd_bosh,start_link,undefined} at <0.1020.0> exit with reason bad record state in ejabberd_bosh:drop_holding_receiver/1 line 701 in context child_terminated
...
```
So at the end, my clients can't associate to my cluster :-(

**Comment by joudinet** [comment]
I found my mistake. I applied 548ef7b without its following fixup: 67fc201.
After patching this patch, no more crash in ejabberd_bosh.
As I haven't seen the failure on mnesia change_table_copy_type since then, may I close this bug and assume it has been somehow fixed by a3083b5?

**Comment by cromain** [comment]
Yes, thous should have been fixed in 18.06
